File: mpmt1.hs
--
--  mpmt1.hs: A stupid simple example of Haskell threading (forkIO)
--
-- License:
--   Apache License, Version 2.0
-- History:
--   * 2024/05/12 v0.1 Initial version
-- Author:
--   Masanori Itoh <masanori.itoh@gmail.com>
-- Usage:
---  $ ghc -threaded  -rtsopts mpmt1.hs -o mpmt1hs
---  $ ./mpmt1hs NUM_CONTEXT DURATION(in sec.) +RTS -Nn (n: number of threads)
---  Note that n of -Nn should be greater than equal NUM_CONTEXT in order to
---  use requested multi cores fully.
-- TODO:
--   * Use Getopt
--   * Implement Process model
import Data.Time.Clock.POSIX (getPOSIXTime)
import System.Environment
import Control.Monad
import Control.Concurrent
import Control.Concurrent.Chan (newChan, writeChan, readChan)
import Text.Printf

t mul = round . (mul *) <$> getPOSIXTime

busyLoop current time_left = do
  when (time_left > 0)
    $ do
      now  <- t 1000
      let elapse = now - current
      busyLoop now  (time_left - elapse)

busyWorker idx duration worker_chan = do
  printf "busyWorker: starting. idx: %d\n"  idx
  now  <- t 1000
  busyLoop now duration
  writeChan worker_chan idx
  printf "busyWorker: expired. idx: %d\n"  idx

main = do

  args <- getArgs
 
  let num_context = if length args >= 1
                    then (read  (args !! 0) :: Int) else 4
  let duration = if length args >= 2
                 then (read  (args !! 1)  :: Int) else 5

  printf "num_context: %d duration: %d (s)\n"  num_context duration

  worker_chan <- newChan

  forM_ [1..num_context] $ \i -> do
    forkIO $ busyWorker i (duration * 1000) worker_chan

  forM_ [1..num_context] $ \i -> do
    ret <- readChan worker_chan
    printf "main: worker exit. idx: %d\n" ret

File: mpmt1.zig
/// mpmt1.zig: A simple example of Zig multi threading
///
/// License:
///   Apache License, Version 2.0
/// History:
///   * 2024/07/08 v0.1 Initial version
/// Author:
///   Masanori Itoh <masanori.itoh@gmail.com>
/// TODO:
///   * Implement process model (if possible)
///   * Use getopt

const std = @import("std");

fn busy_worker(duration: u64) !void {

    std.debug.print("busy_worker_called {d}\n", .{duration});

    const ts_save = std.time.milliTimestamp();

    while (true) {
        const ts = std.time.milliTimestamp();
        if (ts - ts_save > duration) {
            std.debug.print("Expired.\n", .{});
            break;
        }
    }
}

pub fn main() !void {

    var num_context: usize = 3;
    var duration: u64 = 12;

    const allocator = std.heap.page_allocator;

    const args = try std.process.argsAlloc(allocator);
    if (args.len >= 2) {
        num_context = try std.fmt.parseInt(usize, args[1], 10);
    }
    if (args.len >= 3) {
        duration = try std.fmt.parseInt(u64, args[2], 10);
    }

    std.debug.print("num_context: {d} duration: {d}\n", .{num_context, duration});

    // Create busy_worker threads
    var tlist = std.ArrayList(std.Thread).init(allocator);
    defer tlist.deinit();
    for (0 .. num_context) |i| {
        //https://github.com/ziglang/zig/blob/master/lib/std/Thread.zig
        std.debug.print("Creating {d} th thread. \n", .{i});
        const th = try std.Thread.spawn(.{}, busy_worker, .{@as(u64, duration * 1000)});
        try tlist.append(th);
    }
    // Join busy_worker threads
    for (tlist.items) |th| {
        std.Thread.join(th);
    }
}
